packages/base-controller/src/ControllerMessenger.ts

```diff
@@ -513,27 +550,36 @@ export class ControllerMessenger<
    * @param selector - The selector function used to select relevant data from
    * the event payload. The type of the parameters for this selector must match
    * the type of the payload for this event type.
-   * @template E - A type union of Event type strings.
-   * @template V - The selector return value.
+   * @template EventType - A type union of Event type strings.
+   * @template SelectorReturnValue - The selector return value.
    */
-  subscribe<E extends Event['type'], V>(
-    eventType: E,
-    handler: SelectorEventHandler<V>,
-    selector: SelectorFunction<ExtractEventPayload<Event, E>, V>,
+  subscribe<EventType extends Event['type'], SelectorReturnValue>(
+    eventType: EventType,
+    handler: SelectorEventHandler<SelectorReturnValue>,
+    selector: SelectorFunction<
+      ExtractEventPayload<Event, EventType>,
+      SelectorReturnValue
+    >,
   ): void;
 
-  subscribe<E extends Event['type'], V>(
-    eventType: E,
-    handler: ExtractEventHandler<Event, E>,
-    selector?: SelectorFunction<ExtractEventPayload<Event, E>, V>,
+  subscribe<EventType extends Event['type'], SelectorReturnValue>(
+    eventType: EventType,
+    handler: ExtractEventHandler<Event, EventType>,
+    selector?: SelectorFunction<
+      ExtractEventPayload<Event, EventType>,
+      SelectorReturnValue
+    >,
   ): void {
     let subscribers = this.events.get(eventType);
     if (!subscribers) {
       subscribers = new Map();
       this.events.set(eventType, subscribers);
     }
 
-    subscribers.set(handler, selector);
+    subscribers.set(
+      handler,
+      selector as SelectorFunction<unknown[], SelectorReturnValue>,
```

Comment:
I see that this cast was needed because of the change to `EventSubscriptionMap`. I am having trouble understanding why this is an error though.

Reply:
Type casting removed and `EventSubscriptionMap` typing improved here: 7baebd7814ea5f65962cb4316825c50d2488bfd6.

`EventSubscriptionMap` (`this.events` subscription map) was typed as its category supertype, which made it impossible to assign selector function instances with narrower payloads than `unknown[]` to `this.events`.